File: antlr4/CalculadoraPy/EvalVisitor.py
import math
from LabeledExprVisitor import LabeledExprVisitor
from LabeledExprParser import LabeledExprParser


class EvalVisitor(LabeledExprVisitor):
    def __init__(self):
        self.memory = {}
        self.useDegrees = True  # True = trabajar en grados, False = radianes

    # ID '=' expr
    def visitAssign(self, ctx:LabeledExprParser.AssignContext):
        id_ = ctx.ID().getText()
        value = self.visit(ctx.expr())
        self.memory[id_] = value
        return value

    # expr NEWLINE
    def visitPrintExpr(self, ctx:LabeledExprParser.PrintExprContext):
        value = self.visit(ctx.expr())
        print(value)
        return 0.0

    # INT
    def visitInt(self, ctx:LabeledExprParser.IntContext):
        return float(ctx.INT().getText())

    # ID
    def visitId(self, ctx:LabeledExprParser.IdContext):
        id_ = ctx.ID().getText()
        return self.memory.get(id_, 0.0)

    # expr op=('*'|'/') expr
    def visitMulDiv(self, ctx:LabeledExprParser.MulDivContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == LabeledExprParser.MUL:
            return left * right
        return left / right

    # expr op=('+'|'-') expr
    def visitAddSub(self, ctx:LabeledExprParser.AddSubContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == LabeledExprParser.ADD:
            return left + right
        return left - right

    # (expr)
    def visitParens(self, ctx:LabeledExprParser.ParensContext):
        return self.visit(ctx.expr())

    # factorial
    def visitFactorial(self, ctx:LabeledExprParser.FactorialContext):
        value = int(self.visit(ctx.expr()))
        return float(self.factorial(value))

    def factorial(self, n):
        if n < 0:
            raise RuntimeError("Factorial no definido para negativos")
        result = 1
        for i in range(2, n+1):
            result *= i
        return result

    # funciones matemáticas
    def visitFuncCall(self, ctx:LabeledExprParser.FuncCallContext):
        func = ctx.func.text
        value = self.visit(ctx.expr())

        if self.useDegrees and func in ("sin", "cos", "tan"):
            value = math.radians(value)

        if func == "sin": return math.sin(value)
        if func == "cos": return math.cos(value)
        if func == "tan": return math.tan(value)
        if func == "sqrt": return math.sqrt(value)
        if func == "ln": return math.log(value)
        if func == "log": return math.log10(value)
        raise RuntimeError(f"Función desconocida: {func}")

File: antlr4/Calculadora/EvalVisitor.java
import java.util.HashMap;
import java.util.Map;

public class EvalVisitor extends LabeledExprBaseVisitor<Double> {
    Map<String, Double> memory = new HashMap<>();
    /** true = trabajar en grados, false = radianes */
    boolean useDegrees = true;

    /** ID '=' expr NEWLINE */
    @Override
    public Double visitAssign(LabeledExprParser.AssignContext ctx) {
        String id = ctx.ID().getText();
        Double value = visit(ctx.expr());
        memory.put(id, value);
        return value;
    }

    /** expr NEWLINE */
    @Override
    public Double visitPrintExpr(LabeledExprParser.PrintExprContext ctx) {
        Double value = visit(ctx.expr());
        System.out.println(value);
        return 0.0;
    }

    /** INT */
    @Override
    public Double visitInt(LabeledExprParser.IntContext ctx) {
        return Double.valueOf(ctx.INT().getText());
    }

    /** ID */
    @Override
    public Double visitId(LabeledExprParser.IdContext ctx) {
        String id = ctx.ID().getText();
        if (memory.containsKey(id)) return memory.get(id);
        return 0.0;
    }

    /** expr op=('*'|'/') expr */
    @Override
    public Double visitMulDiv(LabeledExprParser.MulDivContext ctx) {
        double left = visit(ctx.expr(0));
        double right = visit(ctx.expr(1));
        if (ctx.op.getType() == LabeledExprParser.MUL) return left * right;
        return left / right;
    }

    /** expr op=('+'|'-') expr */
    @Override
    public Double visitAddSub(LabeledExprParser.AddSubContext ctx) {
        double left = visit(ctx.expr(0));
        double right = visit(ctx.expr(1));
        if (ctx.op.getType() == LabeledExprParser.ADD) return left + right;
        return left - right;
    }

    /** '(' expr ')' */
    @Override
    public Double visitParens(LabeledExprParser.ParensContext ctx) {
        return visit(ctx.expr());
    }

    /** Factorial */
    @Override
    public Double visitFactorial(LabeledExprParser.FactorialContext ctx) {
        int value = visit(ctx.expr()).intValue();
        return (double) factorial(value);
    }

    private int factorial(int n) {
        if (n < 0) throw new RuntimeException("Factorial no definido para negativos");
        int result = 1;
        for (int i = 2; i <= n; i++) result *= i;
        return result;
    }

    /** Funciones matemáticas*/
    @Override
    public Double visitFuncCall(LabeledExprParser.FuncCallContext ctx) {
        String func = ctx.func.getText();
        double value = visit(ctx.expr());

        if (useDegrees && (func.equals("sin") || func.equals("cos") || func.equals("tan"))) {
            value = Math.toRadians(value);
        }

        switch (func) {
            case "sin": return Math.sin(value);
            case "cos": return Math.cos(value);
            case "tan": return Math.tan(value);
            case "sqrt": return Math.sqrt(value);
            case "ln": return Math.log(value);
            case "log": return Math.log10(value);
            default: throw new RuntimeException("Función desconocida: " + func);
        }
    }
}

